crates/pop-cli/tests/contract.rs

```diff
@@ -7,14 +7,52 @@ use pop_contracts::{
 	contracts_node_generator, dry_run_gas_estimate_instantiate, instantiate_smart_contract,
 	run_contracts_node, set_up_deployment, Contract, UpOpts,
 };
-use std::{path::Path, process::Command as Cmd};
+use serde::{Deserialize, Serialize};
+use std::{path::Path, process::Command as Cmd, time::Duration};
 use strum::VariantArray;
+use subxt::{config::DefaultExtrinsicParamsBuilder as Params, tx::Payload, utils::to_hex};
+use subxt_signer::sr25519::dev;
+use tokio::time::sleep;
 use url::Url;
 
+// This struct implements the [`Payload`] trait and is used to submit
+// pre-encoded SCALE call data directly, without the dynamic construction of transactions.
+struct CallData(Vec<u8>);
+impl Payload for CallData {
+	fn encode_call_data_to(
+		&self,
+		_: &subxt::Metadata,
+		out: &mut Vec<u8>,
+	) -> Result<(), subxt::ext::subxt_core::Error> {
+		out.extend_from_slice(&self.0);
+		Ok(())
+	}
+}
+
+// TransactionData has been copied from wallet_integration.rs
+/// Transaction payload to be sent to frontend for signing.
+#[derive(Serialize, Debug)]
+#[cfg_attr(test, derive(Deserialize, Clone))]
+pub struct TransactionData {
+	chain_rpc: String,
+	call_data: Vec<u8>,
+}
+impl TransactionData {
+	pub fn new(chain_rpc: String, call_data: Vec<u8>) -> Self {
+		Self { chain_rpc, call_data }
+	}
+	pub fn call_data(&self) -> Vec<u8> {
+		self.call_data.clone()
+	}
+}
+
 /// Test the contract lifecycle: new, build, up, call
 #[tokio::test]
 async fn contract_lifecycle() -> Result<()> {
 	const DEFAULT_PORT: u16 = 9944;
+	const DEFAULT_ENDPOINT: &str = "ws://127.0.0.1:9944";
+	const WALLET_INT_URI: &str = "http://127.0.0.1:9090";
```

Comment:
Not for this PR, but likely need something to componentize things a little more and let us alter the address for the wallet integration. This port 9090 is going to conflict with some other tests.

Reply:
Agreed!